### ISSUE TITLE
Sanitize shape parameters

### ISSUE THREAD

**Issue by jngrad** [espressomd/espresso#3719]
Currently, shapes accept any geometrical parameter, even those outside the definition domain of their distance calculation function. Incorrect parameters lead to non-differentiable surfaces or NaN values. We need to provide setters that will throw an error when constructing or updating shapes with  incorrect parameters.

To illustrate the issue, here are a few examples:
- `HollowConicalFrustum`: if one of the radii is a negative number, the surface is non-differentiable (it's self-crossing, looks like a light cone)
- `Slitpore`: if the smoothing radii are too large for the pore length or width, the surface is non-differentiable (in fact, `samples/visualization_constraints.py` called with `--slitpore` generates a self-crossing surface, although it's not easy to spot it visually)
- `Stomatocyte`: if `inner_radius < 1`, we calculate the square root of a negative number

It took me already 5 man-hours to figure out why particles from a LJ gas were getting random kicks from certain shapes and why `Stomatocyte` generated NaNs. I'm not the first one to get bitten by this, judging by the high friction coefficient in the thermostat of sample `visualization_constraints.py` to avoid crashes, and the 7:3:1 rule of thumb in the `Stomatocyte` documentation.

**Comment by fweik** [comment]
In this context we should also consider if it wouldn't be easier to provide the tools (like the `Union` thing) to build up more complex shapes from primitives, instead of hard-coding them. I would expect that it is pretty hard to write a test for the `Stomatocyte` shape, that will check that the invariants that you want to enforce are actually kept.

**Comment by jngrad** [comment]
Yeah the `Stomatocyte` is going to be rough. I couldn't find the equations for the expressions in the source code. WolframAlpha helped me in the past with the Gay-Berne code, and it already found a simpler expression for the first variable `d`. I've brute-forced the shape with 1000 gaussian-distributed geometrical parameters, and from the results it seems you can check if parameters are appropriate by simply evaluating the gradient of the distance function along the main axis to see if it's discontinuous. This can be wrapped into a `#ifndef NDEBUG` if it's too slow.

`Stomatocyte` aside, it shouldn't be too difficult to sanitize the parameters for the shapes we provide. This could even be done in a coding day.

Creating a framework to build new shapes from primitives could be an interesting project. We already do it in tutorial 06. Just an idea: for simple shapes we could parse SVG files containing their profile and then extrude them. This would work for shapes with circular symmetry (hollow conical frustum) and linear symmetry (slitpore, wall, rectangular channel). Smoothing could be done in the SVG file using arcs, or added later in espresso using a single fillet radius. Those shapes could then be rendered in OpenGL using primitive shapes. That's basically how sketch-based CAD softwares work.

**Comment by fweik** [comment]
I'm not sure the Stomatocyte has any users, so we shouldn't put too much work into it. If more complex forms could be build up from simpler once, such single-shot stuff wouldn't need to be in the core code, and subsequently be maintained. This is probably not in incredibly good use of our resources...

**Comment by jngrad** [comment]
The issues with `Stomatocyte` are orthogonal to the problem at hand, so let's focus on the bigger picture. Whether we should invest resources into building a CAD framework is an interesting question, and I would be happy to participate to the development if we decide to do so, but we won't port it back to 4.1, so we'll have to address the issue of sanitizing parameters in any case for 4.1.3. The shapes are probably the only classes in espresso where we don't check input values...

**Comment by RudolfWeeber** [comment]
Let’s do it in a “best effort” fashion and add checks for obvious nonsense such as negative radii. Whatever can be achieved in 1-2 hours.

I agree with @fweik, that we should probably ignore stomatocyte. It’s a candidate for removal.

**Comment by fweik** [comment]
`building a CAD framework` sounds very much out of scope for the project, it is probably simpler to just support some common mesh format is supported (and fast), so that existing specialized tools can be used for the actual geometry work...

**Comment by jngrad** [comment]
The `Cylinder` class can be used as an example to re-design the other classes, it's one of the few that has const getters and setters that process the input parameters:
https://github.com/espressomd/espresso/blob/4e578cde72d5f383eb6a88ab20e3849f246714be/src/shapes/include/shapes/Cylinder.hpp#L92-L96

**Comment by jngrad** [comment]
Hmm that's annoying, there is a situation where the correctness of a geometrical parameter depends on the value of another parameter. In the `Rhomboid` shape, the three edge vectors cannot be collinear otherwise the shape is flat, but we cannot know in a setter if the user is done setting the other parameters.

The correct way to address this dependency is to introduce a private member `is_initialized` and call the validation method if it is false inside the `calculate_dist()` method. However this is not a "simple" check anymore, and if the exception is not handled, stderr will be flooded when rasterizing such a shape, e.g. when setting LB boundaries or rasterizing the object in the visualizer. I'd prefer to avoid that.

**Comment by fweik** [comment]
I think it's the easiest way to just make the parameters immutable.

**Comment by jngrad** [comment]
> I think it's the easiest way to just make the parameters immutable.

I agree, that would be far easier. Yet I just found out about a use case for a `Wall` whose `position` attribute is increased slowly over time to compress a polymer ([10.1021/acs.nanolett.9b02224](https://doi.org/10.1021/acs.nanolett.9b02224) section [Simulation](https://pubs.acs.org/doi/10.1021/acs.nanolett.9b02224#sec4_2)).

**Comment by fweik** [comment]
Yes, but you can just replace the whole wall.

**Comment by jngrad** [comment]
`SimplePore` created with `smoothing_radius=0` generates a stack overflow error message in an infinite loop when visualized in the OpenGL visualizer when the GSL Extrusion library is available, because of a division by zero. The cryptic python backtrace did not help in finding the source of the issue. Took me and @KaiSzuttor several man-hours to figure that one out.

**Comment by jngrad** [comment]
offline discussion:
1. We could have immutable shapes. To alter a shape, create a new one. For technical reasons, if we move to a UUID framework with a pseudo-particle attached to shapes, we would need a way to store the new shape at the same UUID as the old one
2. We could make shape parameters immutable except for the scale (to be implemented), position and axis. The shape parameters cannot lead to ill-defined geometries under non-zero homothecy and rotation transformations.

**Comment by RudolfWeeber** [comment]
@jngrad do you know, what is the status, here?

**Comment by jngrad** [comment]
I haven't worked on this ticket since my last post.
